File: cses/02_sorting_searching/1141_playlist/solution.py
import sys


def main():
    # Đọc toàn bộ dữ liệu một lần bằng fast I/O
    data = sys.stdin.buffer.read().split()
    if not data:
        print(0)
        return
    n = int(data[0])
    a = data[1:1 + n]  # giữ nguyên dạng bytes để so sánh, khỏi ép kiểu int

    last = {}   # id -> chỉ số xuất hiện gần nhất
    left = 0    # biên trái của cửa sổ trượt hiện tại
    best = 0
    # Duyệt con trỏ phải; mở rộng cửa sổ [left, right] gồm các id khác nhau
    for right in range(n):
        x = a[right]
        # Nếu id lặp lại và bản cũ còn nằm trong cửa sổ thì nhảy biên trái
        prev = last.get(x)
        if prev is not None and prev >= left:
            left = prev + 1
        last[x] = right
        # Cập nhật độ dài lớn nhất của dãy hợp lệ kết thúc tại right
        length = right - left + 1
        if length > best:
            best = length

    sys.stdout.write(str(best) + "\n")


main()

File: cses/02_sorting_searching/1141_playlist/solution.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::cin >> n;

    // last_position[id] = chỉ số xuất hiện gần nhất của id đó
    std::unordered_map<int, int> last_position;
    last_position.reserve(static_cast<std::size_t>(n) * 2);

    int left = 0;  // biên trái của cửa sổ trượt hiện tại
    int best = 0;

    // Duyệt con trỏ phải; mở rộng cửa sổ [left, right] gồm các id khác nhau
    for (int right = 0; right < n; ++right) {
        int song;
        std::cin >> song;

        // Nếu id lặp lại và bản cũ còn nằm trong cửa sổ thì nhảy biên trái
        const auto it = last_position.find(song);
        if (it != last_position.end() && it->second >= left) {
            left = it->second + 1;
        }
        last_position[song] = right;

        // Cập nhật độ dài lớn nhất của dãy hợp lệ kết thúc tại right
        best = std::max(best, right - left + 1);
    }

    std::cout << best << '\n';
    return 0;
}
